### ISSUE TITLE
Tabset API discussion

### ISSUE THREAD

**Issue by Quinn-L** [ng-bootstrap/ng-bootstrap#594]
Following my pull request #586 has brought in some interesting design discussion that I think should be discussed as a whole.

* `activeId` should support 'banana-in-the-box' two-way binding. It is simply intuitive and no boiler plating required for keeping the user's bound field in-sync when the tab changes via mouse click.
* the `select` method should probably be internal use only, since it's main usage is for user's clicking the tab via mouse and the code tests whether it can change or not.
* `change` output event, at a first glance, appears redundant for preventing users from selecting a tab, since the tab itself has a `disable` field. However, I think if we modify `change` to handle promise/observable callback prevention this will support a more advance case (eg. a modal dialog asking whether the user wants to change tab or not).
* 'banana-in-the-box' can be split out to listen to the post-change event separately, however adding a `post-change` event that provides more in-depth details about the tab change that occurred (eg. from/to id) might also be desired. I personally don't have a use case with this, I can only think of is a user wants to track their user's behavior.

**Comment by pkozlowski-opensource** [comment]
@Quinn-L I've spent a lot of time thinking of the tabset API based on the use-cases we couldn't support in angular-ui/bootstrap. Based on this I believe that we can't take out items from it. Notably you seem to be suggesting that the `select` method could go away in here:

> the select method should probably be internal use only, since it's main usage is for user's clicking the tab via mouse and the code tests whether it can change or not.

But I'm really not sure what are you trying to say here.... 

Then you suggest changes to the `change` output:

> change output event, at a first glance, appears redundant for preventing users from selecting a tab, since the tab itself has a disable field

The thing is that the `disabled` field can block people from navigating **to** a certain tab. But it is no use for preventing people from navigating away from the currently selected tab! So the **`change` output with an event is essential for this use case and can't be removed**.

Now, given that `change` is here to stay I'm really not sure that we gain that much by providing a very similar method (`activeIdChange`) wit a different payload. Yes, it would reduce boilerplate by few characters in _certain_ use cases. But it would also introduce redundant and confusing API.

For now I don't think that the minimal amount of potential "boilerplate" justifies larger and more confusing API surface. I would say that we should wait and see more use-cases to emerge. The current API might be a bit wordy but at least it is consistent and covers all the existing use-cases.

Based on all the above I suggest that we put this change on a back-burner and wait for more use-cases to emerge.

**Comment by Quinn-L** [comment]
Thanks for the response.

Regarding the `select` method, this was @jnizet's suggestion, since the programmer could just set `activeId`, however it would technically bypass any change from/to logic check, so I think you're correct that it should remain as is.

With the `change` method, I did not think about the change _from_ case, which actually makes more sense.

I understand your caution against `activeIdChange`. The 'Select an active tab by id' demo where the tabs weren't in sync was the first thing I noticed when I played with it. I think this does mean the demo page should warn programmers about adding this boiler plate code, otherwise it is 'unreliable' field since it isn't updated when the front-end user interacts with the tabs without manual intervention.

**Comment by pkozlowski-opensource** [comment]
> The 'Select an active tab by id' demo where the tabs weren't in sync was the first thing I noticed when I played with it. I think this does mean the demo page should warn programmers about adding this boiler plate code, otherwise it is 'unreliable' field since it isn't updated when the front-end user interacts with the tabs without manual intervention.

Yes, you are touching upon a very important point here! Actually this is the only reason for which I would consider having 2-way data binding. 

As things stand today the `activeId` property really means "initially active tab". 

In any case there is a lot of valid points in all you are rising and I'm pretty much sure that this will surface in the future. At the same time I don't want to rush to any solution as I don't know what the "perfect" solution is. This is why I prefer to wait and see more use-cases / wait for other ideas.

In the meantime I'm going to push a change to the demo site so it explicitly uses the `select` method.

Thnx for bringing up those issues / thoughts / proposals! Even if we don't change the API today this is important voice in the discussion. Thnx again.
